packages/foundry/contracts/hooks/NftCheckHook.sol

```diff
@@ -0,0 +1,323 @@
+        }
+
+        IERC20[] memory tokens = _vault.getPoolTokens(pool);
+        uint256[] memory accruedFees = new uint256[](tokens.length);
+        hookAdjustedAmountsOutRaw = amountsOutRaw;
+
+        if (exitFeePercentage > 0) {
+            // Charge fees proportional to the `amountOut` of each token.
+            for (uint256 i = 0; i < amountsOutRaw.length; i++) {
+                uint256 exitFee = amountsOutRaw[i].mulDown(exitFeePercentage);
+                accruedFees[i] = exitFee;
+                hookAdjustedAmountsOutRaw[i] -= exitFee;
+
+                emit ExitFeeCharged(pool, tokens[i], exitFee);
+                // Fees don't need to be transferred to the hook, because donation will redeposit them in the Vault.
+                // In effect, we will transfer a reduced amount of tokensOut to the caller, and leave the remainder
+                // in the pool balance.
+            }
+
+            // Donates accrued fees back to LPs
+            _vault.addLiquidity(
+                AddLiquidityParams({
+                    pool: pool,
+                    to: msg.sender, // It would mint BPTs to router, but it's a donation so no BPT is minted
+                    maxAmountsIn: accruedFees, // Donate all accrued fees back to the pool (i.e. to the LPs)
+                    minBptAmountOut: 0, // Donation does not return BPTs, any number above 0 will revert
+                    kind: AddLiquidityKind.DONATION,
+                    userData: bytes("") // User data is not used by donation, so we can set it to an empty string
+                })
+            );
+        }
+
+        return (true, hookAdjustedAmountsOutRaw);
+    }
+
+    // Permissioned functions
+
+    /**
+     * @notice Sets the hook remove liquidity fee percentage, charged on every remove liquidity operation.
+     * @dev This function must be permissioned.
+     */
+    function setExitFeePercentage(uint64 newExitFeePercentage) external onlyOwner {
+        if (newExitFeePercentage > MAX_EXIT_FEE_PERCENTAGE) {
+            revert ExitFeeAboveLimit(newExitFeePercentage, MAX_EXIT_FEE_PERCENTAGE);
+        }
+        exitFeePercentage = newExitFeePercentage;
+
+        emit ExitFeePercentageChanged(address(this), newExitFeePercentage);
+    }
+
+    // New function to update nftContract
+    function setNftContract(address _newNftContract) external onlyOwner {
+        address oldContract = nftContract;
+        nftContract = _newNftContract;
+        emit NftContractUpdated(oldContract, _newNftContract);
+    }
+
+    // New function to update nftId
+    function setNftId(uint256 _newNftId) external onlyOwner {
+        uint256 oldId = nftId;
+        nftId = _newNftId;
+        emit NftIdUpdated(oldId, _newNftId);
+    }
+
+    function recordInitialLiquidity(uint256 token1Amount, uint256 token2Amount) public onlyVault {
+        require(firstDepositor == address(0), "Initial liquidity already recorded");
+        firstDepositor = msg.sender;
```

Comment:
`msg.sender` is the router, it wouldn't be the user. You need to get the sender from a trusted router, as shown here:

https://docs-v3.balancer.fi/build-a-custom-amm/build-an-amm/extend-existing-pool-type-using-hooks.html#implementing-the-swap-fee-logic